hbase-coprocessor/src/main/java/org/apache/omid/transaction/TransactionVisibilityFilter.java

```diff
@@ -0,0 +1,248 @@
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+package org.apache.omid.transaction;
+
+import com.google.common.base.Optional;
+import com.sun.istack.Nullable;
+import org.apache.hadoop.hbase.Cell;
+import org.apache.hadoop.hbase.CellUtil;
+import org.apache.hadoop.hbase.KeyValue;
+import org.apache.hadoop.hbase.client.Get;
+import org.apache.hadoop.hbase.client.Result;
+import org.apache.hadoop.hbase.filter.Filter;
+import org.apache.hadoop.hbase.filter.FilterBase;
+import org.apache.hadoop.hbase.util.Bytes;
+
+import java.io.IOException;
+import java.util.*;
+
+public class TransactionVisibilityFilter extends FilterBase {
+
+    // optional sub-filter to apply to visible cells
+    private final Filter userFilter;
+    private final SnapshotFilterImpl snapshotFilter;
+    private final Map<Long ,Long> shadowCellCache;
+    private final HBaseTransaction hbaseTransaction;
+    private final Map<String, Long> familyDeletionCache;
+
+    public SnapshotFilter getSnapshotFilter() {
+        return snapshotFilter;
+    }
+
+    public TransactionVisibilityFilter(@Nullable Filter cellFilter,
+                                       SnapshotFilterImpl snapshotFilter,
+                                       HBaseTransaction hbaseTransaction) {
+        this.userFilter = cellFilter;
+        this.snapshotFilter = snapshotFilter;
+        shadowCellCache = new HashMap<>();
+        this.hbaseTransaction = hbaseTransaction;
+        familyDeletionCache = new HashMap<String, Long>();
+    }
+
+    @Override
+    public ReturnCode filterKeyValue(Cell v) throws IOException {
+        if (CellUtils.isShadowCell(v)) {
+            Long commitTs =  Bytes.toLong(CellUtil.cloneValue(v));
+            shadowCellCache.put(v.getTimestamp(), commitTs);
+            // Continue getting shadow cells until one of them fits this transaction
+            if (hbaseTransaction.getStartTimestamp() >= commitTs) {
+                return ReturnCode.NEXT_COL;
+            } else {
+                return ReturnCode.SKIP;
+            }
+        } else if (CellUtils.isFamilyDeleteCell(v)) {
+            //Delete is part of this transaction
+            if (snapshotFilter.getTSIfInTransaction(v, hbaseTransaction).isPresent()) {
+                familyDeletionCache.put(Bytes.toString(CellUtil.cloneFamily(v)), v.getTimestamp());
+                return ReturnCode.NEXT_COL;
+            }
+
+            if (shadowCellCache.containsKey(v.getTimestamp()) &&
+                    hbaseTransaction.getStartTimestamp() >= shadowCellCache.get(v.getTimestamp())) {
+                familyDeletionCache.put(Bytes.toString(CellUtil.cloneFamily(v)), shadowCellCache.get(v.getTimestamp()));
```

Comment:
Avoid allocating memory within this call. Either use TreeMap<byte[],Cell> and don't do the copy or use HashMap<ImmutableBytesPtr,Cell>. You can copy/paste ImmutableBytesPtr from Phoenix - it's a wrapper around byte[] that handles equality and hashcode without doing any copying.